File: src/main/java/net/snipsniper/utils/GradientJButton.java
package net.snipsniper.utils;

import javax.swing.*;
import java.awt.*;
import java.awt.font.FontRenderContext;
import java.awt.font.GlyphVector;
import java.awt.geom.AffineTransform;

public class GradientJButton extends JButton {
    private final SSColor color;
    private final String title;

    public GradientJButton(String title, SSColor color) {
        super(title);
        this.color = color;
        this.title = title;
        setContentAreaFilled(false);
    }

    @Override
    public void paint(Graphics g) {
        Graphics2D g2 = (Graphics2D) g;
        g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
        g2.setRenderingHint(RenderingHints.KEY_RENDERING, RenderingHints.VALUE_RENDER_QUALITY);
        g2.setPaint(color.getGradientPaint(getWidth(), getHeight()));
        if(!isEnabled())
            g2.setPaint(Utils.getDisabledColor());
        g2.fillRect(0, 0, getWidth(), getHeight());
        g2.setColor(Color.BLACK);
        g2.drawRect(0, 0, getWidth()-1, getHeight()-1);
        int drawHeight = (int)(getHeight() / 1.5F);
        g2.setFont(new Font(getFont().getFontName(), Font.PLAIN, drawHeight));

        FontRenderContext frc = g2.getFontRenderContext();
        GlyphVector gv = g2.getFont().createGlyphVector(frc, getText());
        Shape shape = gv.getOutline();

        AffineTransform oldTransform = g2.getTransform();
        g2.translate(getWidth()/2 - g2.getFontMetrics().stringWidth(title)/2, getHeight()/2 + drawHeight / 3);
        g2.setStroke(new BasicStroke(2));
        g2.draw(shape);
        g2.setColor(Color.WHITE);
        g2.fill(shape);
        g2.setTransform(oldTransform);

        if(!isEnabled()) {
            g2.setColor(Utils.getDisabledColor());
            g2.fillRect(0, 0, getWidth(), getHeight());
        }

        g2.dispose();
        super.paint(g2);
    }
}

File: src/main/kotlin/net/snipsniper/utils/GradientJButton.kt
package net.snipsniper.utils

import org.capturecoop.cccolorutils.CCColor
import java.awt.*
import javax.swing.JButton

class GradientJButton(val title: String, val color: CCColor) : JButton(title) {
    init {
        isContentAreaFilled = false
    }

    override fun paint(g: Graphics) {
        (g as Graphics2D).also { g2 ->
            g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON)
            g2.setRenderingHint(RenderingHints.KEY_RENDERING, RenderingHints.VALUE_RENDER_QUALITY)
            g2.paint = color.getGradientPaint(width, height)
            if(!isEnabled)
                g2.paint = Utils.getDisabledColor()
            g2.fillRect(0, 0, width, height)
            g2.color = Color.BLACK
            g2.drawRect(0, 0, width - 1, height - 1)
            val drawHeight = (height / 1.5F).toInt()
            g2.font = Font(font.fontName, Font.PLAIN, drawHeight)

            val frc = g2.fontRenderContext
            val gv = g2.font.createGlyphVector(frc, text)
            val shape = gv.outline

            val oldTransform = g2.transform
            g2.translate(width / 2 - g2.fontMetrics.stringWidth(title) / 2, height / 2 + drawHeight / 3)
            g2.stroke = BasicStroke(2F)
            g2.draw(shape)
            g2.color = Color.WHITE
            g2.fill(shape)
            g2.transform = oldTransform

            if(!isEnabled) {
                g2.color = Utils.getDisabledColor()
                g2.fillRect(0, 0, width, height)
            }

            g2.dispose()
            super.paint(g2)
        }
    }
}